File: Quest/Assets/Src/Controller/AIStrategy/AITournamentStrategy2.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class AITournamentStrategy2 : AITournamentStrategy
{
    public Ability[] CardsToPlay(GameState state, Player player)
    {
        int currentBP = player.GetBattlePoints(state);
        if (currentBP >= 50)
            return null; 

        List<AdventureCard> validCards = new List<AdventureCard>();
        List<string> usedCards = new List<string>();

        foreach (AdventureCard card in player.getPlayersCards())
        {
            if (player.CanPlayCardToBoard(card))
            {
                bool used = false;
                for (int i = 0; i < usedCards.Count && !used; i++)
                    if (usedCards[i].Equals(card.getName()))
                        used = true;
                if (!used)
                {
                    usedCards.Add(card.getName());
                    validCards.Add(card);
                }
            }
        }

        validCards.Sort(new ACardComparer(state));

        List<Ability> abilities = new List<Ability>();
        foreach (AdventureCard card in validCards)
        {
            abilities.Add(card.GetAbilities()[0]);
            currentBP += card.getBattlePoints(state);
            if (currentBP >= 50)
                break;
        }

        return abilities.ToArray();
    }

    public bool DoIParticipateInTournament(GameState state, Player player)
    {
        return true;
    }

    private class ACardComparer : IComparer<AdventureCard>
    {
        GameState state;
        public ACardComparer(GameState state)
        {
            this.state = state;
        }

        public int Compare(AdventureCard x, AdventureCard y)
        {
            if (x.getBattlePoints(state) < y.getBattlePoints(state))
                return 1;
            else if (x.getBattlePoints(state) > y.getBattlePoints(state))
                return -1;
            return 0;
        }
    }
}

File: eclipse/Quest/src/controller/AI/AITournamentStrategy2.java
package controller.AI;

import java.util.ArrayList;
import java.util.Comparator;

import model.cards.AdventureCard;
import model.cards.abilities.Ability;
import model.game.GameState;
import model.player.Player;

public class AITournamentStrategy2 implements AITournamentStrategy {

	@Override
	public boolean DoIParticipateInTournament(GameState state, Player player) {
		return true;
	}

	@Override
	public Ability[] CardsToPlay(GameState state, Player player) {
		int currentBP = player.GetBattlePoints(state);
        if (currentBP >= 50)
            return null; 

        ArrayList<AdventureCard> validCards = new ArrayList<>();
        ArrayList<String> usedCards = new ArrayList<>();

        for (AdventureCard card : player.getPlayersCards())
        {
            if (player.CanPlayCardToBoard(card))
            {
                boolean used = false;
                for (int i = 0; i < usedCards.size() && !used; i++)
                    if (usedCards.get(i).equals(card.getName()))
                        used = true;
                if (!used)
                {
                    usedCards.add(card.getName());
                    validCards.add(card);
                }
            }
        }

        validCards.sort(new Comparator<AdventureCard>() {
			@Override
			public int compare(AdventureCard card1, AdventureCard card2) {
				if (card1.getBattlePoints(state) < card2.getBattlePoints(state))
	                return 1;
	            else if (card1.getBattlePoints(state) > card2.getBattlePoints(state))
	                return -1;
	            return 0;
			}
			});

        ArrayList<Ability> abilities = new ArrayList<>();
        for (AdventureCard card : validCards)
        {
            abilities.add(card.GetAbilities()[0]);
            currentBP += card.getBattlePoints(state);
            if (currentBP >= 50)
                break;
        }

        return abilities.toArray(new Ability[abilities.size()]);
	}

}
